OpenRA.Mods.Common/Traits/World/ControlGroups.cs

```diff
@@ -0,0 +1,140 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2021 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System.Collections.Generic;
+using System.Linq;
+using OpenRA.Primitives;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	public class ControlGroupsInfo : TraitInfo
+	{
+		public readonly int GroupsCount = 10;
+
+		public override object Create(ActorInitializer init) { return new ControlGroups(init.World, this); }
+	}
+
+	[TraitLocation(SystemActors.World | SystemActors.EditorWorld)]
+	public class ControlGroups : IControlGroups, ITick, IGameSaveTraitData
+	{
+		readonly World world;
+		public int GroupsCount { get; private set; }
+
+		readonly Cache<int, List<Actor>> controlGroups = new Cache<int, List<Actor>>(_ => new List<Actor>());
```

Comment:
A cache is a bad fit here IMO - we already define that there are `GroupsCount` groups, but the cache allows any arbitrary input. Can we make this a `List<Actor>[]` created with `GroupsCount` entries in the ctor?

Reply:
This was copied from somewhere and I haven't thought about it. I will look in more detail :+1: